docs: add a draft of kuai proposal (#5)
ckb-js/topics: README.md

## Problems to solve

> There are three obvious obstables hindering the development of DApps on CKB.

Author: ```suggestion
There are three obvious obstacles hindering the development of DApps on CKB.
```

### First: separate runtimes

CKB adopts Cell model which is a generalization of UTXO model so only a **function of state verification** is running on-chain, that's the core of a DApp and theoretically that's all of a DApp.

> But a real-word DApp is for users who are not good at mathematics or programming so the logic of the DApp will be exposed by a graphical interface and a set of prescribed actions. The team of a DApp has to 

Author: ```suggestion
But a real-world DApp is for users who are not good at mathematics or programming so the logic of the DApp will be exposed by a graphical interface and a set of prescribed actions. The team of a DApp has to 
```

>   1. re-implement **function of state verification on-chain** off-chain for testing;

Author: ```suggestion
  1. re-implement **function of state verification on-chain** off the chain for testing;
```

  2. define a **state transition equation** following the **function of state verification** for prescribed actions;
  3. define algorithms to get ideal parameters for the **state transition equation**, namely `input cells` from CKB, and implement a server to map kv data into relational data for graphical interface.

A mass of off-chain works are introduced because runtimes of on-chain and off-chain are separate and follow exclusive paradigms.

**function of state verification on-chain** is totally a pure function which returns code_0 for success and code_other for failure, all its arguments are determined so the result is predictable.

> **function of state verification off-chain** is a rehearsal or replay of **function of state verification off-chain**, they share the same logic but the one off-chain may be re-implemented because the one on-chain is based on rust which is hard for most programmers.

Author: ```suggestion
**function of state verification off-chain** is a rehearsal or replay of **function of state verification on-chain**, they share the same logic but the one off-chain may be re-implemented because the one on-chain is based on rust which is hard for most programmers.
```

**state transition equation** is reversed from the **function of state verification**, this is easy to understand and used in ethereum contract, 

```

> // let's define the function of state verification and expect it be be 0

Author: ```suggestion
// let's define the function of state verification and expect it be 0
```

By the support of hardhat, projects will follow the same convention and generate a contract-specific SDK for outside users.

## Kuai Project

`Kuai` project is a framework to solve problems mentioned above, inspired by [actor model](https://www.wikiwand.com/en/Actor_model), [run](https://run.network/), [hardhat](https://hardhat.org/) and other tools popular in the community.

`Kuai` includes a unified paradigm for development, a set of conventions, fine-grind all-in-one devtools with SDKs and a general aggregation service.

### Unified paradigm for development

Two abstractions will be promoted in `Kuai`: `Data storage` and `Data manipulation`

#### Abstraction of data storage

There'll be three basic data models used in `Kuai` based DApps: `Store`, `Contract`, `Token`

##### Store

The basic function of cells in CKB is storing data. But data are meaningless without schema, so the first basic model introduced by `Kuai` is named `Store` which is used to transform plain data into a structured data with schema, similar to [ORM](https://www.wikiwand.com/en/Object-relational_mapping)

> Note that a `Store` model could be a group of cells matching the same `pattern` and working as an entity so it could be regarded as a virutal DApp. Say we have 2 DApps in `School Roster Store` models, each of them consists of many `Student Store` models. And we are going to build a `Sports Meeting` DApp, a new `Sports Meeting Store` model could be created and consists of partial `Student Store` from School A and B respectively, it should work as same as a `School Roster` DApp.

Author: ```suggestion
Note that a `Store` model could be a group of cells matching the same `pattern` and working as an entity, so it could be regarded as a virtual DApp. Say we have 2 DApps in `School Roster Store` models, each of them consists of many `Student Store` models. And we are going to build a `Sports Meeting` DApp, a new `Sports Meeting Store` model could be created and consist of partial `Student Store` from School A and B, respectively, it should work as same as a `School Roster` DApp.
```

#### Abstraction of data manipulation

DApps can read states from each other freely because data/states are arranged uniformly by the abstraction of data storage.

DApps can also communicate with each other freely if data manipulation has been abstracted.

##### Store

A `Store` model should have 7 basic interfaces

- New(pattern): create an `Store` model binding to specified cells, as a DApp located by the pattern, ps: it's a virtual DApp rather than an explicit DApp, but because cells are following the same rule, they can work together as an entity.
- Destroy(): remove state and turn the cells into empty
- Duplicate(): create a `Store` model from an instance
- Sync(blockNumber?): will load and update data from the chain at latest(or specific block) global state
- Get(path): will read value of specified path
- Set(path, value): will set given value at specified path

> - Delete: (path): remove key/value at specified path

Author: ```suggestion
- Delete(path): remove key/value at the specified path
```

##### Contract

A `Contract` model should have 5 more interfaces than `Store` model

- Deploy(): deploy the contract on the chain
- Extend(contract): extends from an existing `Contract` model for overriding
- Upgrade(): upgrade the contract on the chain
- Run(method: string, params: Array<any>): call a method of the contract to update its state
- Link(interfaces: Array<object>): instantiate a contract SDK from interfaces. mainly used to encode a human-readable method call into a message transferred between `Contract` models.

##### Token

A generally used token must have some methods based on `Contract`

- Metadata(tokenId, metadata?): a setter/getter to the metadata of a specific token id including `name`, `symbol`, `tokenURI`, `totalSupply`)
- Mint(tokenId, amount): Mint a token
- Send(from, to, tokenId, amount): Send a token
- Stake(address, dapp, tokenId, amount, locktime?): stake into a dapp and get staking tokens
- Redeem(address, dapp, tokenId, amount): unstake from a dapp and burn a staking token

> - Lock(address, amount, tokenId, locktime): lock tokens, for cases like pre-sale, bounty of a team

Author: ```suggestion
- Lock(address, tokenId, amount, locktime): lock tokens, for cases like pre-sale, bounty of a team
```
